projects/packages/waf/src/class-waf-runtime.php

```diff
@@ -656,109 +674,88 @@ function ( $f ) {
 		$output   = array();
 		$has_only = isset( $only[0] );
 		$has_excl = isset( $excl[0] );
 
-		if ( $has_only ) {
-			foreach ( $only as $prop ) {
-				if ( isset( $source[ $prop ] ) && $this->key_matches( $prop, $only ) ) {
-					$output[ $prop ] = $source[ $prop ];
-				}
+		foreach ( $source as $source_key => $source_val ) {
+			if ( is_array( $source_val ) ) {
+				// if $source_val looks like a tuple from flatten_array(), then use the tuple as the key and value
+				$source_key = $source_val[0];
+				$source_val = $source_val[1];
 			}
-		} else {
-			$output = $source;
-		}
-
-		if ( $has_excl ) {
-			foreach ( array_keys( $output ) as $k ) {
-				if ( $this->key_matches( $k, $excl ) ) {
-					unset( $output[ $k ] );
-				}
+			$filter_match = ( $flags & self::NORMALIZE_ARRAY_MATCH_VALUES ) > 0 ? $source_val : $source_key;
+			// if this key is on the "exclude" list, skip it
+			if ( $has_excl && $this->key_matches( $filter_match, $excl ) ) {
+				continue;
 			}
+			// if this key isn't in our "only" list, then skip it
+			if ( $has_only && ! $this->key_matches( $filter_match, $only ) ) {
+				continue;
+			}
+			// otherwise add this key/value to our output
+			$output[] = array( $source_key, $source_val );
 		}
 
-		if ( $count_only ) {
+		if ( ( $flags & self::NORMALIZE_ARRAY_COUNT ) > 0 ) {
+			// If we've been told to just count the values, then just count them.
 			$results[] = array(
-				'name'   => $name,
+				'name'   => (string) $name,
 				'value'  => count( $output ),
 				'source' => '&' . $name,
 			);
 		} else {
-			foreach ( $output as $tk => $tv ) {
-				if ( is_array( $tv ) ) {
-					// flatten it so we get all the values considered
-					$flat_values = $this->array_flatten( $tv );
-					foreach ( $flat_values as $fv ) {
-						$results[] = array(
-							// force names to strings
-							// we don't care about the nested keys here, just the overall variable name
-							'name'   => '' . $tk,
-							'value'  => $fv,
-							'source' => "$name:$tk",
-						);
-					}
-				} else {
-					$results[] = array(
-						// force names to strings
-						'name'   => '' . $tk,
-						'value'  => $tv,
-						'source' => "$name:$tk",
-					);
-				}
+			foreach ( $output as list( $item_name, $item_value ) ) {
+				$results[] = array(
+					'name'   => (string) $item_name,
+					'value'  => $item_value,
+					'source' => "$name:$item_name",
+				);
 			}
 		}
 
 		return $results;
 	}
 
 	/**
-	 * Basic array flatten with array_merge; no-op on non-array targets.
+	 * Given an array of tuples - probably from flatten_array() - return a new array
+	 * consisting of only the first value (the key name) from each tuple.
 	 *
-	 * @param array $source Array to flatten.
-	 * @return array The flattened array.
+	 * @param array{0:string, 1:scalar}[] $flat_array An array of tuples.
+	 * @return string[]
 	 */
-	private function array_flatten( $source ) {
```

Comment:
Replaced this method with `flatten_array()` utility function (in `src/functions.php` below) that does a better job of transforming a nested array into something that ModSecurity rules are expecting.